dnf5/commands/check-upgrade/check-upgrade.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "check-upgrade.hpp"
+
+#include "utils/bgettext/bgettext-mark-domain.h"
+
+#include <dnf5/shared_options.hpp>
+#include <libdnf-cli/output/changelogs.hpp>
+#include <libdnf-cli/output/package_list_sections.hpp>
+#include <libdnf/conf/const.hpp>
+#include <libdnf/rpm/package_query.hpp>
+
+#include <iostream>
+
+namespace dnf5 {
+
+using namespace libdnf::cli;
+
+void CheckUpgradeCommand::set_parent_command() {
+    auto * arg_parser_parent_cmd = get_session().get_argument_parser().get_root_command();
+    auto * arg_parser_this_cmd = get_argument_parser_command();
+    arg_parser_parent_cmd->register_command(arg_parser_this_cmd);
+    arg_parser_parent_cmd->get_group("software_management_commands").register_argument(arg_parser_this_cmd);
+}
+
+void CheckUpgradeCommand::set_argument_parser() {
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description("Check for available package upgrades");
+
+    changelogs = dynamic_cast<libdnf::OptionBool *>(
```

Comment:
Probably nothing serious, but is really the `dynamic_cast` necessary if we would separate the `unique_ptr` creation and raw pointer assignments?

Reply:
I was thinking about not using the `add_init_value` return value at all, because it only returns the argument that was passed in. The overall usage of that function seems kinda weird to me, because it seems only moving the ownership of options to the parser and nothing else. And just because of that there are then additional runtime casting etc... But anyway, you're right that it's used like that on many more places, so it would be better to do some systemic change about that.